l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -0,0 +1,638 @@
+        _;
+    }
+
+    /// @notice Checks that the message sender is the bridgehub or zkSync Era Diamond Proxy.
+    modifier onlyBridgehubOrEra(uint256 _chainId) {
+        require(
+            msg.sender == address(bridgehub) || (_chainId == ERA_CHAIN_ID && msg.sender == ERA_DIAMOND_PROXY),
+            "L1SharedBridge: not bridgehub or era chain"
+        );
+        _;
+    }
+
+    /// @notice Checks that the message sender is the legacy bridge.
+    modifier onlyLegacyBridge() {
+        require(msg.sender == address(legacyBridge), "ShB not legacy bridge");
+        _;
+    }
+
+    /// @dev Contract is expected to be used as proxy implementation.
+    /// @dev Initialize the implementation to prevent Parity hack.
+    constructor(
+        address _l1WethAddress,
+        IBridgehub _bridgehub,
+        IL1ERC20Bridge _legacyBridge
+    ) reentrancyGuardInitializer {
+        _disableInitializers();
+        l1WethAddress = _l1WethAddress;
+        bridgehub = _bridgehub;
+        legacyBridge = _legacyBridge;
+    }
+
+    /// @dev Initializes a contract bridge for later use. Expected to be used in the proxy
+    /// @param _owner Address which can change L2 token implementation and upgrade the bridge
+    /// implementation. The owner is the Governor and separate from the ProxyAdmin from now on, so that the Governor can call the bridge.
+    function initialize(
+        address _owner,
+        uint256 _eraFirstPostUpgradeBatch
+    ) external reentrancyGuardInitializer initializer {
+        require(_owner != address(0), "ShB owner 0");
+        _transferOwnership(_owner);
+
+        eraFirstPostUpgradeBatch = _eraFirstPostUpgradeBatch;
+        l2BridgeAddress[ERA_CHAIN_ID] = ERA_ERC20_BRIDGE_ADDRESS;
+    }
+
+    /// @dev Initializes the l2Bridge address by governance for a specific chain.
+    function initializeChainGovernance(uint256 _chainId, address _l2BridgeAddress) external onlyOwner {
+        l2BridgeAddress[_chainId] = _l2BridgeAddress;
+    }
+
+    /// @notice Allows bridgehub to acquire mintValue for L1->L2 transactions.
+    /// @dev If the corresponding L2 transaction fails, refunds are issued to a refund recipient on L2.
+    function bridgehubDepositBaseToken(
+        uint256 _chainId,
+        address _prevMsgSender,
+        address _l1Token,
+        uint256 _amount
+    ) external payable virtual onlyBridgehubOrEra(_chainId) {
+        if (_l1Token == ETH_TOKEN_ADDRESS) {
+            require(msg.value == _amount, "L1SharedBridge: msg.value not equal to amount");
+        } else {
+            // The Bridgehub also checks this, but we want to be sure
+            require(msg.value == 0, "ShB m.v > 0 b d.it");
+
+            uint256 amount = _depositFunds(_prevMsgSender, IERC20(_l1Token), _amount); // note if _prevMsgSender is this contract, this will return 0. This does not happen.
+            require(amount == _amount, "3T"); // The token has non-standard transfer logic
```

Comment:
Ineffective check for non-standard transfer logic, case in point: Upgradable tokens. They can introduce non-standard transfer logic after passing this check.
 
 Also, **maybe** [flash-mintable tokens](https://github.com/d-xo/weird-erc20?tab=readme-ov-file#flash-mintable-tokens) can also cause some issues, probably something like a guaranteed revert

Reply:
I'm closing non-standard transfer logic threads, they are are out of scope for now.